---
title: "Working with RangedSummarizedExperiments with DNA Methylation Data"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{identifying_tmrs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 12,
  fig.height = 6.75
)
```

```{r setup, message = F}
library(methodical)
library(dplyr)
```

## Introduction

Most functions from `methodical` take as input RangedSumamrizedExperiment objects 
with methylation data. If there are many samples, there can be many millions
or, in the case of WGBS data, even billions of data points within a DNA methylation 
dataset. It can be unfeasible to load all this data into memory at once. This problem 
can be overcome by using DelayedArrays backed by HDF5 files, enabling data to be 
read into memory only as needed. Methodical provides a suite of functions 
for working with such DNA methylation RangedSumamrizedExperiments, including 
functions to extract methylation values for sites overlapping genomic regions of 
interest `GRanges`, to randomly sample methylation sites, to liftover the
methylation sites from one genome build to another and to mask methylation sites
overlapping certain genomic regions, e.g. repeats. 

Here we demonstrate this different functionality using a dataset downloaded
from TumourMethData.

```{r, eval=TRUE}
## CHANGE TO USE TUMOURMETHDATA
library(TumourMethData)
tcga_wgbs_hg38 = TumourMethData::download_meth_dataset("tcga_wgbs_hg38")
#tcga_wgbs_hg38 = HDF5Array::loadHDF5SummarizedExperiment("~/wgbs/tcga_wgbs/tcga_wgbs_hg38/")
```

## Extracting methylation data from methylation RangedSummarizedExperiments

We'll first demonstrate how to extract methylation data for individual CpG sites 
in tcga_wgbs_hg38 overlapping a supplied GRanges object with 
`extractGRangesMethSiteValues`, using the APC gene body as an example.


```{r, eval=TRUE}
# Create a GRanges with the hg38 genomic coordinates around the start site of 
# the APC gene, including 2 KB upstream of its designated start in Ensembl
apc_gr <- GRanges("chr5:112705518-112716517:+")

# Extract methylation values for CpG sites overlapping apc gene body
apc_cpg_methylation <- extractGRangesMethSiteValues(
  meth_rse = tcga_wgbs_hg38, genomic_regions = apc_gr)

# View the first few rows and columns of the result. 
# extractGRangesMethSiteValues returns a row for each methylation site and a 
# separate column for each sample where row names give the coordinates of the 
# methylation sites in character format. 
apc_cpg_methylation[1:6, 1:6]
```

Next we'll show how to summarize the methylation of CpGs over regions of 
interest with `summarizeRegionMethylation`, using CpG islands as an example. 
`summarizeRegionMethylation` processes the the supplied genomic regions in 
chunks so that the methylation data for CpG sites overlapping the regions of 
interest is not all loaded into memory at once. The parameter 
`max_sites_per_chunk` controls the approximate number of CpG sites maximally 
read into memory at once and defaults to `floor(62500000/ncol(meth_rse)`. 
Several chunks can be processed in parallel using BiocParallel via the 
`BPPARAM` argument which takes a BiocParallelParam object. The number of workers
indicated by BiocParallelParam determines the number of chunks that will be 
processed in parallel. Some experimentation may be needed to find the optimal 
choices for `max_sites_per_chunk` and the number of workers in terms of speed 
and memory usage. 

```{r, eval=TRUE}
# Load CpG islands annotation for hg38
cpg_island_annotation <- annotatr::build_annotations(genome = "hg38", annotation = "hg38_cpgs")
names(cpg_island_annotation) <- cpg_island_annotation$id
cpg_island_annotation <- GRangesList(split(cpg_island_annotation, cpg_island_annotation$type))

cpg_island_methylation <- summarizeRegionMethylation(
  meth_rse = tcga_wgbs_hg38, genomic_regions = cpg_island_annotation$hg38_cpg_islands, 
  BPPARAM = BiocParallel::MulticoreParam(workers = 4), summary_function = colMeans)

# Print a few rows for the first few samples of the result
cpg_island_methylation[1000:1006, 1:6]
```

# We can also summarize methylation values overlapping regions of interest

# Load CpG islands annotation for hg38
cpg_island_annotation <- annotatr::build_annotations(genome = "hg38", annotation = "hg38_cpgs")
names(cpg_island_annotation) <- cpg_island_annotation$id
cpg_island_annotation <- GRangesList(split(cpg_island_annotation, cpg_island_annotation$type))

# Calculate the mean methylation of CpGs in each CpG island for each sample
cpg_island_methylation <- summarizeRegionMethylation(
  meth_rse = tcga_wgbs_hg38, genomic_regions = cpg_island_annotation$hg38_cpg_islands, 
  BPPARAM = BiocParallel::bpparam(), summary_function = colMeans)


# Plot the methylation values along the APC gene body for one colon tumour sample.
plotMethSiteValues(apc_cpg_methylation, column_name = "TCGA_A2_A0YG_01", 
  ylabel = "Methylation Level")

# We can also plot the difference in methylation values between samples,
# such as between colon tumour and its matching normal sample, by creating a 
# data.frame with a column with these values.
apc_cpg_methylation_change <- data.frame(meth_change = 
  apc_cpg_methylation$TCGA_AA_3518_01 - apc_cpg_methylation$TCGA_AA_3518_11,
  row.names = row.names(apc_cpg_methylation))

# Plot mean difference in methylation between colon tumour and its matching normal sample.
apc_coad_meth_change_plot <- plotMethSiteValues(apc_cpg_methylation_change, 
  column_name = "meth_change", ylabel = "Methylation Change in Colon Cancer")
print(apc_coad_meth_change_plot)
  

annotateMethSitePlot(meth_site_plot = apc_coad_meth_change_plot, 
  annotation_grl = cpg_island_annotation, annotation_plot_proportion = 0.3)

```

## Creation of methylation RSE from array files

Generally, files from methylation array files will have one column with the 
name of the probe used to measure methylation at a particular CpG site and
another column with the value associated with this probe. 
`makeMethRSEFromArrayFiles` takes a GRanges object for the `probe_ranges` argument 
to match the names of these probe to genomic coordinates when creating the meth RSE. 
This should be a GRanges object with one metadata object called "name" which 
gives the names of the probes and matches the names in the array files. 

One such GRanges object, `infinium_450k_probe_granges_hg19` for mapping probes 
from the Illumina HumanMethylation450 array to genomic coordinates of the hg19 
human genome build is provided as part of the package. If mapping the probes to 
a genome build besides hg19 or using a different methylation array, another 
similar GRanges object with the coordinates of probes and a metadata column
with their names should be provided.

We'll take a look at one of the methylation files we just downloaded and at
`infinium_450k_probe_granges_hg19`.

```{r, eval=TRUE}
# View the top of one of the COAD methylation array files
system2("head", meth_array_files[1])

# Load infinium_450k_probe_granges_hg19 and view the first few ranges
data("infinium_450k_probe_granges_hg19")
head(infinium_450k_probe_granges_hg19)
```

Now we'll create the meth RSE for the four files we downloaded. It is assumed
that the first column in the methylation array files is the name of the probes
and the second is the values associated with them but this can be changed using
the `probe_name_column` and `beta_value_column` arguments. 

Files can be processed in parallel using BiocParallel with the `BPPARAM` argument,
which defaults to `BiocParallel::bpparam()`. 

We'll save the HDF5 directory within the temporary directory using the `hdf5_dir`, 
but this should be saved somewhere else if the meth RSE will be reused. 

```{r, eval=TRUE}

# Create a data.frame with sample metadata associated with the array files
sample_metadata = data.frame(cancer_type = rep("COAD",  4),
  sample_type = rep("primary_tumor", 4), row.names = names(meth_array_files))

# Create meth RSE for COAD
coad_meth_rse = makeMethRSEFromArrayFiles(
  array_files = meth_array_files, 
  probe_name_column = 1, beta_value_column = 2, 
  probe_ranges = infinium_450k_probe_granges_hg19, 
  sample_metadata = sample_metadata, BPPARAM = BiocParallel::bpparam(),
  hdf5_dir = paste(tempdir(), "coad_meth_array_rse", sep = "/"))

# Display coad_meth_rse
print(coad_meth_rse)
```

## SessionInfo
```{r}
sessionInfo()
```